### DISCUSSION TITLE
belongsTo on pivot table

### DISCUSSION THREAD

**Discussion by caiokawasaki** [laravel/framework#49710] [Q&A]
Yesterday I spent a few hours trying to resolve this, I ended up with a solution but I didn't really like it.

I started developing a multi-tenancy application and came across an issue when resolving user relations.

Each user can only belong to one `Tenant`, when he creates his account he is automatically attached to the `Tenant` related to the domain. When any user is loaded, `Tenant` is automatically loaded so that the system can know the user's role, the only issue is that I couldn't bring in just one `Tenant`, I need to bring them all in and then filter.

These are my models:

```php
class User extends Authenticatable
{
    protected $fillable = [
        'first_name',
        'last_email',
        'password',
    ];

    protected $hidden = [
        'password',
        'remember_token',
    ];

    protected $casts = [
        'email_verified_at' => 'datetime',
        'password' => 'hashed',
    ];

    protected $with = [
        'tenants',
    ];


    public function tenants(): BelongsToMany
    {
        return $this->belongsToMany(Tenant::class, 'tenant_user')
            ->using(TenantUser::class)
            ->withPivot(['role']);
    }
}
```

```php
class Tenant extends Model
{
    protected $fillable = [
        'name',
        'domain',
    ];


    public function users(): BelongsToMany
    {
        return $this->belongsToMany(User::class, 'tenant_user')
            ->using(TenantUser::class)
            ->withPivot(['role']);
    }
}
```

My pivot class:

```php
class TenantUser extends Pivot
{
    protected $table = 'tenant_user';

    protected $fillable = [
        'role'
    ];

    protected $hidden = [
        'tenant_id',
        'user_id',
    ];


    public function tenant(): BelongsTo
    {
        return $this->belongsTo(Tenant::class, 'tenant_id');
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class, 'user_id');
    }
}
```

My solution was the following:

```php
class User extends Authenticatable
{
    protected $fillable = [
        'first_name',
        'last_email',
        'password',
    ];

    protected $hidden = [
        'password',
        'remember_token',
    ];

    protected $casts = [
        'email_verified_at' => 'datetime',
        'password' => 'hashed',
    ];

    protected $with = [
        'tenants',
    ];

    protected $appends = [
        'role',
    ];
    
    
    public function tenants(): BelongsToMany
    {
        return $this->belongsToMany(Tenant::class, 'tenant_user')
            ->using(TenantUser::class)
            ->withPivot(['role']);
    }


    protected function role(): Attribute
    {
        return new Attribute(
            get: fn() => $this->tenants->first()->pivot->role
        );
    }
}
```

I would like to do something similar to this:

```php
class User extends Authenticatable
{
    protected $fillable = [
        'first_name',
        'last_email',
        'password',
    ];

    protected $hidden = [
        'password',
        'remember_token',
    ];

    protected $casts = [
        'email_verified_at' => 'datetime',
        'password' => 'hashed',
    ];

    protected $with = [
        'tenant',
    ];

    protected $appends = [
        'role',
    ];


    public function tenant(): BelongsToMany
    {
        return $this->belongsTo(Tenant::class, 'tenant_user') // belongsTo is only accepted when in the same table
            ->using(TenantUser::class)
            ->withPivot(['role']);
    }


    protected function role(): Attribute
    {
        return new Attribute(
            get: fn() => $this->tenant->pivot->role
        );
    }
}
```

Is this possible?

### Edit

I achieved something similar to what I want with the following relationship:

```php
class User extends Authenticatable
{
	public function tenant()
    {
        return $this->hasOneThrough(
            Tenant::class,
            TenantUser::class,
            'tenant_id',
            'id',
            'id',
            'user_id'
        );
    }
}
```

However, this way I cannot access the other columns of the pivot table.

**Comment by haleyngonadi** [comment]
Since a user has just one tenant, should the users table have a tenent_id?

**Reply by caiokawasaki** [reply]
I want to use a separate table because user will have configurations related to the tenant, I don't want to keep it on the users table

**Reply by macropay-solutions** [reply]
@caiokawasaki You could create a new relation class that extends BelongsToMany and in it you should overwrite

     /**
     * Get the results of the relationship.
     *
     * @return mixed
     */
    public function getResults()
    {
        return ! is_null($this->parent->{$this->parentKey})
                ? $this->get()
                : $this->related->newCollection();
    }

so you get first result. Maybe 

    return parent::getResults()->first();

Also unique on user_id from pivot.

in belongs to:

    /**
     * Get the results of the relationship.
     *
     * @return mixed
     */
    public function getResults()
    {
        if (is_null($this->child->{$this->foreignKey})) {
            return $this->getDefaultFor($this->parent);
        }

        return $this->query->first() ?: $this->getDefaultFor($this->parent);
    }
